Added Support For Big Endian UTF-32

NoMoreFood opened the pull request:
# PR Summary

Added Support For Big Endian UTF-32 to common encoding enumerations/classes.

## PR Context

Big Endian UTF-32 is a somewhat common encoding and was supported under Windows PowerShell.  This pull adds support for it and also addresses a small bug that resulted in a test being effectively skipped.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/PowerShell/PowerShell/issues/11645
- **Testing - New and feature**
    - [] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 commented on src/System.Management.Automation/utils/EncodingUtils.cs:
```diff
@@ -25,13 +26,17 @@ internal static class EncodingConversion
             };
 
+        // BigEndianUTF32 encoding is possible, but requires creation
+        internal static Encoding BigEndianUTF32Encoding = new UTF32Encoding(bigEndian: true, byteOrderMark: true);
```
Should this field be `readonly`?

NoMoreFood replied:
Originally I had just copied this from another function where it was defined like this, but I updated this commit to just embed it in the dictionary.  Since we're only using it once, no need to generate a separate object.

vexx32 commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -45,42 +45,36 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
+    $availableEncodings = 
+        @([System.Text.Encoding]::ASCII,
+          [System.Text.Encoding]::BigEndianUnicode,
+          (New-Object System.Text.UTF32Encoding($true,$true)),
```
Couple things here, I think..

1. This should be accompanied with a comment indicating why it's different from the below UTF32 encoding, since the code itself doesn't really make it clear unless you're familiar with the constructor.
2. Might just be my own preference, but I'd probably also put it closer to the other UTF32 value in the list.
3. It's probably preferable to use `[System.Text.UTF32Encoding]::new(...)` rather than `New-Object` since the cmdlet is generally a good bit slower.

The syntax here can also be simplified; array subexpressions accept newlines as item separators anyway, and semicolons aren't required in PS for line endings:

```powershell
$availableEncodings = @(
    [System.Text.Encoding]::ASCII
    [System.Text.Encoding]::BigEndianUnicode
    # ..... (etc)
)
```

NoMoreFood replied:
Addressed latest commit.  I left the ordering the same since it lines up with the other, similar enumerations in the source (which are actually alphabetically ordered).

vexx32 commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -45,42 +45,36 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
-                for($i = 0;$i -lt $observedBytes.Count; $i++) {
-                    $observedBytes[$i] | Should -Be $expectedBytes[$i]
-                }
+        $encodingName = $encoding.EncodingName
```
This comment should be updated, since we're changing how this is done here.

NoMoreFood replied:
The comment purely applicable to how the encoding was being parsed so, to your point, I just removed it altogether.

NoMoreFood committed:
Added Support For Big Endian UTF-32

- Added support for big endian UTF-32 and common classes.
- Updated relevant tests for big endian UTF-32.
- Fixed test that was attempting to enumerate encoding types but was enumerating an empty set.

vexx32 commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -45,42 +45,36 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
+        $BOM = $encoding.GetPreamble()
+        $TXT = $encoding.GetBytes($asciiString)
+        $CR  = $encoding.GetBytes($asciiCR)
+        $expectedBytes = .{ $BOM; $TXT; $CR }
```
Not sure why the original test was written this way, this is a bit expensive just to essentially put an array together. Better to use a subexpression here as well:

```suggestion
        $expectedBytes = @( $BOM; $TXT; $CR )
```

You can use line breaks instead of semicolons if you prefer.

NoMoreFood replied:
Made these adjustments in the latest commit.

vexx32 commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -45,42 +45,36 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
+        It $msg {
+            $observedBytes.Count | Should -Be $expectedBytes.Count
+            for($i = 0;$i -lt $observedBytes.Count; $i++) {
+                $observedBytes[$i] | Should -Be $expectedBytes[$i]
```
... huh. I guess it's a little out of scope since we're updating an existing test, but the way this is written is _very_ screwy. This should really be refactored to use `-TestCases` instead of a manual loop here, and any setup that has to happen outside the `It` block should go into a `BeforeAll {}` block... Most of it can really happen inside the `It` in any case.

The `for` loop is also unnecessary; Pester is perfectly capable of verifying that all items in two arrays match.

If you'd rather not do a more extensive refactor in this PR, give me a ping when this is merged and I'll come back and refactor the test more thoroughly. 🙂

NoMoreFood replied:
Other changes have been merged.  I'm heading to bed so feel refactor away ;-)

vexx32 reviewed (changes requested):
Nice work!

There's a few things we can do a bit better here, but overall this looks good. Thanks for adding this! 💖 😊

NoMoreFood committed:
Added Support For Big Endian UTF-32: Review Changes

- Addressed review comments in test scripts.

JamesWTruher reviewed (approved):
LGTM

adityapatwardhan commented:
@NoMoreFood Thank you for your contribution!
